Add a command-line interface

mtpereira opened the pull request.

mtpereira committed:
Move screenshot logic to its own package

Add a screenshooter package that implements taking screenshots of a
website using chromdp.

mtpereira committed:
Add a command-line interface

In order to use the Screenshooter, a CLI was added. It allows for
pointing to a URL and taking a screenshot of it.

mtpereira commented on screenshooter.go:
```diff
@@ -0,0 +1,66 @@
+		return err
+	}
+
+	err = ioutil.WriteFile(fmt.Sprintf("%s/%s.png", ss.SavePath, filename), buf, 0644)
```
Will handle already existing files later.

mtpereira commented on cmd/hugo-post-preview/hugo-post-preview.go:
```diff
@@ -0,0 +1,52 @@
+package main
```
Is this the idiomatic way to build binaries and CLI tools on Go? In particular, what package should they be in and is this directory tree ok?

mtpereira replied:
I've followed the blog post at https://ieftimov.com/post/golang-package-multiple-binaries/ and renamed this file to `main.go` accordingly.

jacoelho replied:
cmd/<binary>/main.go

example
https://github.com/kubernetes/kubernetes/tree/master/cmd

mtpereira replied:
It's fine then, awesome! :smile:

mtpereira commented on screenshooter.go:
```diff
@@ -0,0 +1,66 @@
+
+// New initializes and returns a Screenshooter.
+func New(config ...func(*Screenshooter)) *Screenshooter {
+	ctx, cancel := chromedp.NewContext(context.Background(), chromedp.WithDebugf(log.Printf))
```
Debugging should be configurable from the CLI interface.

mtpereira committed:
Improve Screenshooter's configuration

Provide a set of functions for configuring the currently available
parameters for the Screenshooter. This is a cleaner interface for
passing configuration to the object.

jacoelho commented on screenshooter.go:
```diff
@@ -47,8 +71,8 @@ func elementScreenshot(urlstr, sel string, res *[]byte) chromedp.Tasks {
 // Take a screenshot of the given post.
 func (ss *Screenshooter) Take(postURL url.URL, element string, filename string) error {
```
I would change function signature:
```
Take(ctx context.Context, postURL url.URL, element string, w io.Writer) error 
```

let te caller decide which context and destination to use without opinions

errm replied:
+1 on taking a `io.Writer` rather than a filename... it makes testing much simpler too!

mtpereira replied:
Oh yes, this should definitely be an `io.Writer`, thanks! :bowing_man: 

I'll keep the context out of the signature for now, since it is but an implementation detail of the `chromedp` library. If it starts to be more useful (for instance, to enable more flexible configuration), I'll add it to the signature. Sticking with the YAGNI mindset on this one. :stuck_out_tongue:

mtpereira replied:
Done! Looking good, if I can say so myself! :laughing:

jacoelho commented on screenshooter.go:
```diff
@@ -16,22 +16,46 @@ type Screenshooter struct {
-		ctx:    ctx,
-		cancel: cancel,
-	}
+	ss := Screenshooter{}
```
nitpick

```
var ss Screenshoter
```

to keep consistent when declaring nil slices

errm commented on screenshooter.go:
```diff
@@ -57,7 +81,7 @@ func (ss *Screenshooter) Take(postURL url.URL, element string, filename string)
 	}
 
-	err = ioutil.WriteFile(fmt.Sprintf("%s/%s.png", ss.SavePath, filename), buf, 0644)
+	err = ioutil.WriteFile(fmt.Sprintf("%s/%s.png", ss.savePath, filename), buf, 0644)
```
If `ss.savePath` path is empty then the file would written to `/filename` which is possibly not desirable.

mtpereira replied:
> If `ss.savePath` path is empty then the file would written to `/filename` which is possibly not desirable.

Uhm yeah, I might want to have some sort of default and/or error handling here. Will do.

mtpereira replied:
> not sure how much you care, you could use https://golang.org/pkg/path/filepath/ to support cross platform

Not a lot, to be honest, at least not for now. I didn't know that pkg existed, so I've learnt something today, thanks! :smile:

errm reviewed (commented):
nit: the object should be called `Screenshotter`

I like the change to the functional config, it defiantly makes the API cleaner!

mtpereira committed:
Simplify writing the screenshot to a file

Use a `io.Writer` on `*Screenshooter.Take` so that the pkg user can use
whatever they desire to write the screenshot. This also simplifies the
pkg since it no longer needs to handle file names and paths.

The CLI now defaults to writing to `./post.png` whenever the `-filename`
is not provided.

mtpereira committed:
Fix the name of the Screenshotter

Proper English, you should use it.

mtpereira commented:
Thanks for all the feedback, this looks good now! :bowing_man:
